packages/flutter/lib/src/rendering/object.dart

```diff
@@ -158,16 +164,31 @@ class PaintingContext extends ClipContext {
         return true;
       }());
     }
-    child._layer.offset = offset;
-    _appendLayer(child._layer);
+    compositeLayerAtOffset(child._layer, offset);
   }
 
-  void _appendLayer(Layer layer) {
+  /// Adds an already detached layer.
+  @protected
+  void appendLayer(Layer layer) {
     assert(!_isRecording);
-    layer.remove();
+    assert(!layer.attached);
     _containerLayer.append(layer);
   }
 
+  /// Composites a [layer] at the specified [offset].
+  ///
+  /// Subclasses that need to avoid modifying existing layers must reimplement
+  /// this method using an alternate mechanism to composite the [layer] at the
+  /// specified [offset].
+  @protected
+  void compositeLayerAtOffset(OffsetLayer layer, Offset offset) {
+    appendLayer(
+      layer
+        ..offset = offset
+        ..remove(),
```

Comment:
we shouldn't be calling remove() unless it's attached... can we assert that it's attached here? this seems like an odd contract.

Reply:
this is identical to what was previously happening although hidden inside the _appendLayer helper method. 
`Layer.remove` gracefully handles detached layers which is required because some of the layers past here are attached to a parent and some are not.  Perhaps the documentation for remove should be updated?
```dart
 /// Removes this layer from its parent layer's child list.
  @mustCallSuper
  void remove() {
    parent?._removeChild(this);
  }
```

Reply:
fair enough. that wasn't a fantastic use of cascades. updated documentation for `remove` and `compositeLayerAtOffset`.